File: scripts/lint/lint-commits.py
#!/usr/bin/env python3
"""Lint commit messages with the project-local commitlint and message policy.

Resolves the commitlint binary, collects the commits in the current range, and
runs each message through commitlint and validate-commit-message.py. Replaces
the former lint-commits.sh while preserving its environment contract and exit
codes (0 success, 1 lint failure, 2 missing commitlint).

Dependabot pull requests are exempt: a pull request opened by the
`dependabot[bot]` author is skipped entirely, and commits authored by
`dependabot[bot]` are skipped only in push ranges, because Dependabot
generates its own branch names and commit messages. Human branches keep every
rule unchanged.
"""

from __future__ import annotations

import importlib.util
import os
import subprocess
import sys
from pathlib import Path

SCRIPT_DIR = Path(__file__).resolve().parent

DEPENDABOT_AUTHOR = "dependabot[bot]"


def is_dependabot_pull_request() -> bool:
    """Return whether the current pull request author is Dependabot."""
    return os.environ.get("PR_AUTHOR", "") == DEPENDABOT_AUTHOR


def _load_commit_message_validator():
    """Import validate-commit-message.py despite the hyphenated filename."""
    path = SCRIPT_DIR / "validate-commit-message.py"
    spec = importlib.util.spec_from_file_location("validate_commit_message", path)
    assert spec and spec.loader
    validator = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(validator)
    return validator


VALIDATE_COMMIT_MESSAGE = _load_commit_message_validator()


def resolve_commitlint(root, runner=subprocess) -> str:
    """Return the commitlint binary path, exiting 2 when none is available."""
    override = os.environ.get("COMMITLINT_BIN")
    if override:
        return override
    candidate = Path(root) / ".mise" / "bin" / "commitlint"
    if os.access(candidate, os.X_OK):
        return str(candidate)
    print(
        "commitlint is not installed for this project. Run: mise run setup:commitlint",
        file=sys.stderr,
    )
    raise SystemExit(2)


def collect_commits(base, head, runner=subprocess) -> list[str]:
    """Return commit SHAs in a base..head range, or from a single head."""
    argv = ["git", "rev-list", "--no-merges"]
    argv.append(f"{base}..{head}" if base else head)
    result = runner.run(argv, text=True, capture_output=True)
    return [line for line in result.stdout.splitlines() if line]


def commit_author(commit, runner=subprocess) -> str:
    """Return the author name recorded on a commit."""
    return runner.check_output(
        ["git", "show", "-s", "--format=%an", commit], text=True
    ).strip()


def is_dependabot_commit(commit, runner=subprocess) -> bool:
    """Return whether a commit was authored by Dependabot."""
    return commit_author(commit, runner) == DEPENDABOT_AUTHOR


def message_for(commit, runner=subprocess) -> str:
    """Return the full commit message body for a commit.

    Trailing newlines are stripped just as Bash command substitution did in the
    former lint-commits.sh, so a single-line message is not treated as having a
    spurious empty body line.
    """
    return runner.check_output(
        ["git", "show", "-s", "--format=%B", commit], text=True
    ).rstrip("\n")


def lint_message(message, commitlint_bin, runner=subprocess) -> int:
    """Lint one message, returning 0 on success or the failing status."""
    result = runner.run([commitlint_bin, "lint"], input=message, text=True)
    if result.returncode != 0:
        return result.returncode
    errors = VALIDATE_COMMIT_MESSAGE.validate(message)
    if errors:
        print("error: " + "; ".join(errors), file=sys.stderr)
        return 1
    print("Commit message shape is valid.")
    return 0


def main(runner=subprocess) -> int:
    root = runner.check_output(
        ["git", "rev-parse", "--show-toplevel"], text=True
    ).strip()
    if is_dependabot_pull_request():
        print("Skipping commit lint for a Dependabot pull request.")
        return 0
    commitlint_bin = resolve_commitlint(root, runner)
    pr_context = bool(os.environ.get("PR_BASE_SHA") and os.environ.get("PR_HEAD_SHA"))
    if pr_context:
        commits = collect_commits(
            os.environ["PR_BASE_SHA"], os.environ["PR_HEAD_SHA"], runner
        )
    else:
        before = os.environ.get("PUSH_BEFORE_SHA", "")
        after = os.environ.get("GITHUB_SHA") or "HEAD"
        if before and not all(character == "0" for character in before):
            commits = collect_commits(before, after, runner)
        else:
            commits = collect_commits(None, after, runner)
    if not commits:
        print("No commits to lint.")
        return 0
    for commit in commits:
        print(f"Checking commit {commit}")
        if not pr_context and is_dependabot_commit(commit, runner):
            print(f"Skipping Dependabot-authored commit {commit}")
            continue
        message = message_for(commit, runner)
        status = lint_message(message, commitlint_bin, runner)
        if status != 0:
            return status
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: internal/commitlint/lintcommits.go
package commitlint

import (
	"context"
	"fmt"
	"io"
	"os"
	"strings"

	"github.com/hidekitux/skills/internal/provider"
)

const dependabotAuthor = "dependabot[bot]"

// commandRunner abstracts subprocess execution so the linter is testable
// without a real Git repository. runValue writes input to stdin (empty for
// none) and returns combined output and the process exit code.
type commandRunner interface {
	runValue(input string, name string, args ...string) (string, int)
}

// providerRunner runs the linter's commands through the provider ports: git
// through the Git port and the pinned commitlint binary through the Tool port.
type providerRunner struct {
	git  provider.Git
	tool provider.Tool
}

// ExecRunner returns the real subprocess-backed command runner used in
// production command entrypoints.
func ExecRunner() commandRunner {
	runner := provider.OSRunner{}
	return providerRunner{git: provider.NewGit(runner), tool: provider.NewTool(runner)}
}

func (r providerRunner) runValue(input, name string, args ...string) (string, int) {
	ctx := context.Background()
	if name == "git" && input == "" {
		result, _ := r.git.Output(ctx, "", args...)
		return result.Combined, result.ExitCode
	}
	result, _ := r.tool.Invoke(ctx, provider.Command{Name: name, Args: args, Stdin: input})
	return result.Combined, result.ExitCode
}

func isDependabotPullRequest() bool {
	return os.Getenv("PR_AUTHOR") == dependabotAuthor
}

// resolveCommitlint returns the commitlint binary path, or an error carrying
// exit code 2 when none is installed.
func resolveCommitlint(root string, runner commandRunner) (string, int) {
	if override := os.Getenv("COMMITLINT_BIN"); override != "" {
		return override, 0
	}
	candidate := root + "/.mise/bin/commitlint"
	if info, err := os.Stat(candidate); err == nil && !info.IsDir() && info.Mode()&0o111 != 0 {
		return candidate, 0
	}
	return "", 2
}

func collectCommits(base, head string, runner commandRunner) []string {
	argv := []string{"rev-list", "--no-merges"}
	if base != "" {
		argv = append(argv, base+".."+head)
	} else {
		argv = append(argv, head)
	}
	stdout, _ := runner.runValue("", "git", argv...)
	var commits []string
	for _, line := range strings.Split(stdout, "\n") {
		if line != "" {
			commits = append(commits, line)
		}
	}
	return commits
}

func commitAuthor(commit string, runner commandRunner) string {
	stdout, _ := runner.runValue("", "git", "show", "-s", "--format=%an", commit)
	return strings.TrimSpace(stdout)
}

func isDependabotCommit(commit string, runner commandRunner) bool {
	return commitAuthor(commit, runner) == dependabotAuthor
}

func messageFor(commit string, runner commandRunner) string {
	stdout, _ := runner.runValue("", "git", "show", "-s", "--format=%B", commit)
	return strings.TrimSuffix(stdout, "\n")
}

func lintMessage(message, commitlintBin string, runner commandRunner, out, errOut io.Writer) int {
	stdout, code := runner.runValue(message, commitlintBin, "lint")
	if code != 0 {
		fmt.Fprint(errOut, stdout)
		return code
	}
	errors := ValidateMessage(message)
	if len(errors) > 0 {
		fmt.Fprintf(errOut, "error: %s\n", strings.Join(errors, "; "))
		return 1
	}
	fmt.Fprintln(out, "Commit message shape is valid.")
	return 0
}

// LintCommits drives commitlint over the commits in the current push or pull
// request range, returning the process exit code.
func LintCommits(runner commandRunner, out, errOut io.Writer) int {
	root, _ := runner.runValue("", "git", "rev-parse", "--show-toplevel")
	root = strings.TrimSpace(root)
	if isDependabotPullRequest() {
		fmt.Fprintln(out, "Skipping commit lint for a Dependabot pull request.")
		return 0
	}
	commitlintBin, code := resolveCommitlint(root, runner)
	if code != 0 {
		fmt.Fprintln(errOut, "commitlint is not installed for this project. Run: mise run setup:commitlint")
		return 2
	}
	prContext := os.Getenv("PR_BASE_SHA") != "" && os.Getenv("PR_HEAD_SHA") != ""
	var commits []string
	if prContext {
		commits = collectCommits(os.Getenv("PR_BASE_SHA"), os.Getenv("PR_HEAD_SHA"), runner)
	} else {
		before := os.Getenv("PUSH_BEFORE_SHA")
		after := os.Getenv("GITHUB_SHA")
		if after == "" {
			after = "HEAD"
		}
		if before != "" && !isAllZeros(before) {
			commits = collectCommits(before, after, runner)
		} else {
			commits = collectCommits("", after, runner)
		}
	}
	if len(commits) == 0 {
		fmt.Fprintln(out, "No commits to lint.")
		return 0
	}
	for _, commit := range commits {
		fmt.Fprintf(out, "Checking commit %s\n", commit)
		if !prContext && isDependabotCommit(commit, runner) {
			fmt.Fprintf(out, "Skipping Dependabot-authored commit %s\n", commit)
			continue
		}
		message := messageFor(commit, runner)
		if status := lintMessage(message, commitlintBin, runner, out, errOut); status != 0 {
			return status
		}
	}
	return 0
}

func isAllZeros(sha string) bool {
	for _, r := range sha {
		if r != '0' {
			return false
		}
	}
	return true
}
